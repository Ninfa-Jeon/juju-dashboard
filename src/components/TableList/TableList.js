import React from "react";
import { Link } from "react-router-dom";
import { useSelector } from "react-redux";

import MainTable from "../MainTable/MainTable";

/**
  Returns the model info and statuses in the proper format for the table data.
  @param {Object} state The application state.
  @returns {Object} The formatted table data.
*/
function generateModelTableData(state) {
  const models = state.juju.models;
  const modelInfo = state.juju.modelInfo;
  return Object.keys(models).map(modelUUID => {
    const modelListData = models[modelUUID];
    const modelInfoData = modelInfo[modelUUID];
    const modelStatus = state.juju.modelStatuses[modelUUID];
    const modelDetailsPath = `/models/${modelListData.name}`;
    const generateModelDetailsLink = () => {
      const owner = modelListData.ownerTag;
      const user = state.root.controllerConnection.info.user.identity;
      if (owner === user) {
        return <Link to={modelDetailsPath}>{modelListData.name}</Link>;
      }
      // If the owner isn't the logged in user then we need to use the
      // fully qualified path name.
      const sharedModelDetailsPath = `/models/${owner.replace("user-", "")}/${
        modelListData.name
      }`;
      return <Link to={sharedModelDetailsPath}>{modelListData.name}</Link>;
    };
    return {
      columns: [
        { content: generateModelDetailsLink() },
        { content: modelListData.ownerTag.split("@")[0].replace("user-", "") },
        { content: getStatusValue(modelStatus, "summary") },
        { content: getStatusValue(modelStatus, "cloudTag") },
        { content: getStatusValue(modelStatus, "region") },
        { content: getStatusValue(modelInfoData, "cloudCredentialTag") },
        // We're not currently able to get the controller name from the API.
        // so display the controller UUID instead.
        { content: getStatusValue(modelInfoData, "controllerUuid") },
        // We're not currently able to get a last-accessed or updated from JAAS.
        { content: getStatusValue(modelInfoData, "status.since") }
      ]
    };
  });
}

/**
  Used to fetch the values from status as it won't be defined when the
  modelInfo data is.
  @param {Object|undefined} status The status for the model.
  @param {String} key The key to fetch.
  @returns {String} The computed value for the requested field if defined, or
    an empty string.
*/
function getStatusValue(status, key) {
  let returnValue = "";
  if (typeof status === "object" && status !== null) {
    switch (key) {
      case "summary":
        const applicationKeys = Object.keys(status.applications);
        const applicationCount = applicationKeys.length;
        const machineCount = Object.keys(status.machines).length;
        const unitCount = applicationKeys.reduce(
          (prev, key) =>
            prev + Object.keys(status.applications[key].units).length,
          0
        );
        returnValue = applicationCount + "/" + machineCount + "/" + unitCount;
        break;
      case "cloudTag":
        returnValue = status.model.cloudTag.replace("cloud-", "");
        break;
      case "region":
        returnValue = status.model.region;
        break;
      case "cloudCredentialTag":
        returnValue = status.cloudCredentialTag
          .split("cloudcred-")[1]
          .split("@")[1]
          .split("_")[1];
        break;
      case "controllerUuid":
        returnValue = status.controllerUuid.split("-")[0] + "...";
        break;
      case "status.since":
        returnValue = status.status.since.split("T")[0];
        break;
      default:
        console.log(`unsupported status value key: ${key}`);
        break;
    }
  }
  return returnValue;
}

const modelTableHeaders = [
  { content: "Name", sortKey: "name" },
  { content: "Owner", sortKey: "owner" },
  { content: "Apps/Machines/Units", sortKey: "summary" },
  { content: "Cloud", sortKey: "cloud" },
  { content: "Region", sortKey: "region" },
  { content: "Credential", sortKey: "credential" },
  { content: "Controller", sortKey: "controller" },
  { content: "Last Updated", sortKey: "last-updated" }
];

function TableList() {
  const modelData = useSelector(generateModelTableData);
  return (
    <MainTable
      className={"u-table-layout--auto"}
      headers={modelTableHeaders}
      rows={modelData}
    />
  );
}

export default TableList;
