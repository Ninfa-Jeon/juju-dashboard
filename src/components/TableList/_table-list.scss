@import "vanilla-framework/scss/vanilla";

.table-list_error-message {
  color: $color-negative;
  font-size: 0.875rem;
  text-overflow: ellipsis;
  white-space: nowrap;
}
