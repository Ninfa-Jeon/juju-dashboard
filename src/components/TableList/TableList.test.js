import React from "react";
import { MemoryRouter } from "react-router";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";

import TableList from "./TableList";

import dataDump from "../../testing/complete-redux-store-dump";

const mockStore = configureStore([]);

describe("TableList", () => {
  it("by default, renders with all table headers and no data", () => {
    const store = mockStore({
      root: {},
      juju: {
        models: {},
        modelData: {},
        modelInfo: {},
        modelStatuses: {}
      }
    });
    const wrapper = mount(
      <Provider store={store}>
        <TableList />
      </Provider>
    );
    expect(wrapper.find("th")).toMatchSnapshot();
    // Should be empty
    expect(wrapper.find("tbody")).toMatchSnapshot();
  });

  // XXX Skipped until the test data can be updated.
  // https://github.com/canonical-web-and-design/jaas-dashboard/issues/115
  it.skip("displays all data from redux store", () => {
    const store = mockStore(dataDump);
    const wrapper = mount(
      <MemoryRouter>
        <Provider store={store}>
          <TableList />
        </Provider>
      </MemoryRouter>
    );
    expect(wrapper.find(TableList)).toMatchSnapshot();
  });
});
