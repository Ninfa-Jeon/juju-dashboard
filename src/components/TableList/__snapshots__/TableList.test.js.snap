// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableList by default, renders with all table headers and no data 1`] = `
Array [
  <th
    aria-sort="none"
    onClick={[Function]}
    role="columnheader"
  >
    Name
  </th>,
  <th
    aria-sort="none"
    onClick={[Function]}
    role="columnheader"
  >
    Owner
  </th>,
  <th
    aria-sort="none"
    onClick={[Function]}
    role="columnheader"
  >
    Summary
  </th>,
  <th
    aria-sort="none"
    onClick={[Function]}
    role="columnheader"
  >
    Cloud
  </th>,
  <th
    aria-sort="none"
    onClick={[Function]}
    role="columnheader"
  >
    Region
  </th>,
  <th
    aria-sort="none"
    onClick={[Function]}
    role="columnheader"
  >
    Credential
  </th>,
  <th
    aria-sort="none"
    onClick={[Function]}
    role="columnheader"
  >
    Controller
  </th>,
  <th
    aria-sort="none"
    onClick={[Function]}
    role="columnheader"
  >
    Last Updated
  </th>,
]
`;

exports[`TableList by default, renders with all table headers and no data 2`] = `<tbody />`;

exports[`TableList displays all data from redux store 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "mc2fht",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "mc2fht",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <TableList>
        <MainTable
          headers={
            Array [
              Object {
                "content": "Name",
                "sortKey": "name",
              },
              Object {
                "content": "Owner",
                "sortKey": "owner",
              },
              Object {
                "content": "Summary",
                "sortKey": "summary",
              },
              Object {
                "content": "Cloud",
                "sortKey": "cloud",
              },
              Object {
                "content": "Region",
                "sortKey": "region",
              },
              Object {
                "content": "Credential",
                "sortKey": "credential",
              },
              Object {
                "content": "Controller",
                "sortKey": "controller",
              },
              Object {
                "content": "Last Updated",
                "sortKey": "last-updated",
              },
            ]
          }
          rows={
            Array [
              Object {
                "columns": Array [
                  Object {
                    "content": <Link
                      to="/models/group-test"
                    >
                      group-test
                    </Link>,
                  },
                  Object {
                    "content": "mrdata",
                  },
                  Object {
                    "content": "9 applications, 0 machines, 0 units",
                  },
                  Object {
                    "content": "google",
                  },
                  Object {
                    "content": "us-central1",
                  },
                  Object {
                    "content": "google_mrdata@external_admin",
                  },
                  Object {
                    "content": "a030329a-941f-4760-8fcd-3063b41a04e7",
                  },
                  Object {
                    "content": "unavailable",
                  },
                ],
              },
              Object {
                "columns": Array [
                  Object {
                    "content": <Link
                      to="/models/test1"
                    >
                      test1
                    </Link>,
                  },
                  Object {
                    "content": "mrdata",
                  },
                  Object {
                    "content": "0 applications, 0 machines, 0 units",
                  },
                  Object {
                    "content": "google",
                  },
                  Object {
                    "content": "us-east1",
                  },
                  Object {
                    "content": "google_mrdata@external_juju",
                  },
                  Object {
                    "content": "a030329a-941f-4760-8fcd-3063b41a04e7",
                  },
                  Object {
                    "content": "unavailable",
                  },
                ],
              },
              Object {
                "columns": Array [
                  Object {
                    "content": <Link
                      to="/models/test2"
                    >
                      test2
                    </Link>,
                  },
                  Object {
                    "content": "mrdata",
                  },
                  Object {
                    "content": "0 applications, 0 machines, 0 units",
                  },
                  Object {
                    "content": "google",
                  },
                  Object {
                    "content": "us-east1",
                  },
                  Object {
                    "content": "google_mrdata@external_juju",
                  },
                  Object {
                    "content": "a030329a-941f-4760-8fcd-3063b41a04e7",
                  },
                  Object {
                    "content": "unavailable",
                  },
                ],
              },
              Object {
                "columns": Array [
                  Object {
                    "content": <Link
                      to="/models/test3"
                    >
                      test3
                    </Link>,
                  },
                  Object {
                    "content": "mrdata",
                  },
                  Object {
                    "content": "0 applications, 0 machines, 0 units",
                  },
                  Object {
                    "content": "google",
                  },
                  Object {
                    "content": "us-east1",
                  },
                  Object {
                    "content": "google_mrdata@external_juju",
                  },
                  Object {
                    "content": "a030329a-941f-4760-8fcd-3063b41a04e7",
                  },
                  Object {
                    "content": "unavailable",
                  },
                ],
              },
              Object {
                "columns": Array [
                  Object {
                    "content": <Link
                      to="/models/juju-all-the-things"
                    >
                      juju-all-the-things
                    </Link>,
                  },
                  Object {
                    "content": "otheruser",
                  },
                  Object {
                    "content": "3 applications, 3 machines, 3 units",
                  },
                  Object {
                    "content": "google",
                  },
                  Object {
                    "content": "us-east1",
                  },
                  Object {
                    "content": "google_otheruser@external_google-bigbot",
                  },
                  Object {
                    "content": "a030329a-941f-4760-8fcd-3063b41a04e7",
                  },
                  Object {
                    "content": "unavailable",
                  },
                ],
              },
              Object {
                "columns": Array [
                  Object {
                    "content": <Link
                      to="/models/frontend-ci"
                    >
                      frontend-ci
                    </Link>,
                  },
                  Object {
                    "content": "space-man",
                  },
                  Object {
                    "content": "1 applications, 1 machines, 1 units",
                  },
                  Object {
                    "content": "google",
                  },
                  Object {
                    "content": "us-east1",
                  },
                  Object {
                    "content": "google_space-man@external_gce",
                  },
                  Object {
                    "content": "a030329a-941f-4760-8fcd-3063b41a04e7",
                  },
                  Object {
                    "content": "unavailable",
                  },
                ],
              },
              Object {
                "columns": Array [
                  Object {
                    "content": <Link
                      to="/models/backend-ci"
                    >
                      backend-ci
                    </Link>,
                  },
                  Object {
                    "content": "space-man",
                  },
                  Object {
                    "content": "3 applications, 3 machines, 3 units",
                  },
                  Object {
                    "content": "google",
                  },
                  Object {
                    "content": "us-east1",
                  },
                  Object {
                    "content": "google_space-man@external_gce",
                  },
                  Object {
                    "content": "a030329a-941f-4760-8fcd-3063b41a04e7",
                  },
                  Object {
                    "content": "unavailable",
                  },
                ],
              },
              Object {
                "columns": Array [
                  Object {
                    "content": <Link
                      to="/models/data-aggregator"
                    >
                      data-aggregator
                    </Link>,
                  },
                  Object {
                    "content": "space-man",
                  },
                  Object {
                    "content": "2 applications, 2 machines, 2 units",
                  },
                  Object {
                    "content": "google",
                  },
                  Object {
                    "content": "us-east1",
                  },
                  Object {
                    "content": "google_space-man@external_gce",
                  },
                  Object {
                    "content": "a030329a-941f-4760-8fcd-3063b41a04e7",
                  },
                  Object {
                    "content": "unavailable",
                  },
                ],
              },
            ]
          }
          sortable={true}
        >
          <Table
            sortable={true}
          >
            <table
              className="p-table--sortable"
              role="grid"
            >
              <thead>
                <TableRow>
                  <tr
                    role="row"
                  >
                    <TableHeader
                      key="0"
                      onClick={[Function]}
                      sort="none"
                    >
                      <th
                        aria-sort="none"
                        onClick={[Function]}
                        role="columnheader"
                      >
                        Name
                      </th>
                    </TableHeader>
                    <TableHeader
                      key="1"
                      onClick={[Function]}
                      sort="none"
                    >
                      <th
                        aria-sort="none"
                        onClick={[Function]}
                        role="columnheader"
                      >
                        Owner
                      </th>
                    </TableHeader>
                    <TableHeader
                      key="2"
                      onClick={[Function]}
                      sort="none"
                    >
                      <th
                        aria-sort="none"
                        onClick={[Function]}
                        role="columnheader"
                      >
                        Summary
                      </th>
                    </TableHeader>
                    <TableHeader
                      key="3"
                      onClick={[Function]}
                      sort="none"
                    >
                      <th
                        aria-sort="none"
                        onClick={[Function]}
                        role="columnheader"
                      >
                        Cloud
                      </th>
                    </TableHeader>
                    <TableHeader
                      key="4"
                      onClick={[Function]}
                      sort="none"
                    >
                      <th
                        aria-sort="none"
                        onClick={[Function]}
                        role="columnheader"
                      >
                        Region
                      </th>
                    </TableHeader>
                    <TableHeader
                      key="5"
                      onClick={[Function]}
                      sort="none"
                    >
                      <th
                        aria-sort="none"
                        onClick={[Function]}
                        role="columnheader"
                      >
                        Credential
                      </th>
                    </TableHeader>
                    <TableHeader
                      key="6"
                      onClick={[Function]}
                      sort="none"
                    >
                      <th
                        aria-sort="none"
                        onClick={[Function]}
                        role="columnheader"
                      >
                        Controller
                      </th>
                    </TableHeader>
                    <TableHeader
                      key="7"
                      onClick={[Function]}
                      sort="none"
                    >
                      <th
                        aria-sort="none"
                        onClick={[Function]}
                        role="columnheader"
                      >
                        Last Updated
                      </th>
                    </TableHeader>
                  </tr>
                </TableRow>
              </thead>
              <tbody>
                <TableRow
                  key="0"
                >
                  <tr
                    role="row"
                  >
                    <TableCell
                      key="0"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        <Link
                          to="/models/group-test"
                        >
                          <a
                            href="/models/group-test"
                            onClick={[Function]}
                          >
                            group-test
                          </a>
                        </Link>
                      </td>
                    </TableCell>
                    <TableCell
                      key="1"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        mrdata
                      </td>
                    </TableCell>
                    <TableCell
                      key="2"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        9 applications, 0 machines, 0 units
                      </td>
                    </TableCell>
                    <TableCell
                      key="3"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google
                      </td>
                    </TableCell>
                    <TableCell
                      key="4"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        us-central1
                      </td>
                    </TableCell>
                    <TableCell
                      key="5"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google_mrdata@external_admin
                      </td>
                    </TableCell>
                    <TableCell
                      key="6"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        a030329a-941f-4760-8fcd-3063b41a04e7
                      </td>
                    </TableCell>
                    <TableCell
                      key="7"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        unavailable
                      </td>
                    </TableCell>
                  </tr>
                </TableRow>
                <TableRow
                  key="1"
                >
                  <tr
                    role="row"
                  >
                    <TableCell
                      key="0"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        <Link
                          to="/models/test1"
                        >
                          <a
                            href="/models/test1"
                            onClick={[Function]}
                          >
                            test1
                          </a>
                        </Link>
                      </td>
                    </TableCell>
                    <TableCell
                      key="1"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        mrdata
                      </td>
                    </TableCell>
                    <TableCell
                      key="2"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        0 applications, 0 machines, 0 units
                      </td>
                    </TableCell>
                    <TableCell
                      key="3"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google
                      </td>
                    </TableCell>
                    <TableCell
                      key="4"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        us-east1
                      </td>
                    </TableCell>
                    <TableCell
                      key="5"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google_mrdata@external_juju
                      </td>
                    </TableCell>
                    <TableCell
                      key="6"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        a030329a-941f-4760-8fcd-3063b41a04e7
                      </td>
                    </TableCell>
                    <TableCell
                      key="7"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        unavailable
                      </td>
                    </TableCell>
                  </tr>
                </TableRow>
                <TableRow
                  key="2"
                >
                  <tr
                    role="row"
                  >
                    <TableCell
                      key="0"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        <Link
                          to="/models/test2"
                        >
                          <a
                            href="/models/test2"
                            onClick={[Function]}
                          >
                            test2
                          </a>
                        </Link>
                      </td>
                    </TableCell>
                    <TableCell
                      key="1"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        mrdata
                      </td>
                    </TableCell>
                    <TableCell
                      key="2"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        0 applications, 0 machines, 0 units
                      </td>
                    </TableCell>
                    <TableCell
                      key="3"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google
                      </td>
                    </TableCell>
                    <TableCell
                      key="4"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        us-east1
                      </td>
                    </TableCell>
                    <TableCell
                      key="5"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google_mrdata@external_juju
                      </td>
                    </TableCell>
                    <TableCell
                      key="6"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        a030329a-941f-4760-8fcd-3063b41a04e7
                      </td>
                    </TableCell>
                    <TableCell
                      key="7"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        unavailable
                      </td>
                    </TableCell>
                  </tr>
                </TableRow>
                <TableRow
                  key="3"
                >
                  <tr
                    role="row"
                  >
                    <TableCell
                      key="0"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        <Link
                          to="/models/test3"
                        >
                          <a
                            href="/models/test3"
                            onClick={[Function]}
                          >
                            test3
                          </a>
                        </Link>
                      </td>
                    </TableCell>
                    <TableCell
                      key="1"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        mrdata
                      </td>
                    </TableCell>
                    <TableCell
                      key="2"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        0 applications, 0 machines, 0 units
                      </td>
                    </TableCell>
                    <TableCell
                      key="3"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google
                      </td>
                    </TableCell>
                    <TableCell
                      key="4"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        us-east1
                      </td>
                    </TableCell>
                    <TableCell
                      key="5"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google_mrdata@external_juju
                      </td>
                    </TableCell>
                    <TableCell
                      key="6"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        a030329a-941f-4760-8fcd-3063b41a04e7
                      </td>
                    </TableCell>
                    <TableCell
                      key="7"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        unavailable
                      </td>
                    </TableCell>
                  </tr>
                </TableRow>
                <TableRow
                  key="4"
                >
                  <tr
                    role="row"
                  >
                    <TableCell
                      key="0"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        <Link
                          to="/models/juju-all-the-things"
                        >
                          <a
                            href="/models/juju-all-the-things"
                            onClick={[Function]}
                          >
                            juju-all-the-things
                          </a>
                        </Link>
                      </td>
                    </TableCell>
                    <TableCell
                      key="1"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        otheruser
                      </td>
                    </TableCell>
                    <TableCell
                      key="2"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        3 applications, 3 machines, 3 units
                      </td>
                    </TableCell>
                    <TableCell
                      key="3"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google
                      </td>
                    </TableCell>
                    <TableCell
                      key="4"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        us-east1
                      </td>
                    </TableCell>
                    <TableCell
                      key="5"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google_otheruser@external_google-bigbot
                      </td>
                    </TableCell>
                    <TableCell
                      key="6"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        a030329a-941f-4760-8fcd-3063b41a04e7
                      </td>
                    </TableCell>
                    <TableCell
                      key="7"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        unavailable
                      </td>
                    </TableCell>
                  </tr>
                </TableRow>
                <TableRow
                  key="5"
                >
                  <tr
                    role="row"
                  >
                    <TableCell
                      key="0"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        <Link
                          to="/models/frontend-ci"
                        >
                          <a
                            href="/models/frontend-ci"
                            onClick={[Function]}
                          >
                            frontend-ci
                          </a>
                        </Link>
                      </td>
                    </TableCell>
                    <TableCell
                      key="1"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        space-man
                      </td>
                    </TableCell>
                    <TableCell
                      key="2"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        1 applications, 1 machines, 1 units
                      </td>
                    </TableCell>
                    <TableCell
                      key="3"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google
                      </td>
                    </TableCell>
                    <TableCell
                      key="4"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        us-east1
                      </td>
                    </TableCell>
                    <TableCell
                      key="5"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google_space-man@external_gce
                      </td>
                    </TableCell>
                    <TableCell
                      key="6"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        a030329a-941f-4760-8fcd-3063b41a04e7
                      </td>
                    </TableCell>
                    <TableCell
                      key="7"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        unavailable
                      </td>
                    </TableCell>
                  </tr>
                </TableRow>
                <TableRow
                  key="6"
                >
                  <tr
                    role="row"
                  >
                    <TableCell
                      key="0"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        <Link
                          to="/models/backend-ci"
                        >
                          <a
                            href="/models/backend-ci"
                            onClick={[Function]}
                          >
                            backend-ci
                          </a>
                        </Link>
                      </td>
                    </TableCell>
                    <TableCell
                      key="1"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        space-man
                      </td>
                    </TableCell>
                    <TableCell
                      key="2"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        3 applications, 3 machines, 3 units
                      </td>
                    </TableCell>
                    <TableCell
                      key="3"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google
                      </td>
                    </TableCell>
                    <TableCell
                      key="4"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        us-east1
                      </td>
                    </TableCell>
                    <TableCell
                      key="5"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google_space-man@external_gce
                      </td>
                    </TableCell>
                    <TableCell
                      key="6"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        a030329a-941f-4760-8fcd-3063b41a04e7
                      </td>
                    </TableCell>
                    <TableCell
                      key="7"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        unavailable
                      </td>
                    </TableCell>
                  </tr>
                </TableRow>
                <TableRow
                  key="7"
                >
                  <tr
                    role="row"
                  >
                    <TableCell
                      key="0"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        <Link
                          to="/models/data-aggregator"
                        >
                          <a
                            href="/models/data-aggregator"
                            onClick={[Function]}
                          >
                            data-aggregator
                          </a>
                        </Link>
                      </td>
                    </TableCell>
                    <TableCell
                      key="1"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        space-man
                      </td>
                    </TableCell>
                    <TableCell
                      key="2"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        2 applications, 2 machines, 2 units
                      </td>
                    </TableCell>
                    <TableCell
                      key="3"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google
                      </td>
                    </TableCell>
                    <TableCell
                      key="4"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        us-east1
                      </td>
                    </TableCell>
                    <TableCell
                      key="5"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        google_space-man@external_gce
                      </td>
                    </TableCell>
                    <TableCell
                      key="6"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        a030329a-941f-4760-8fcd-3063b41a04e7
                      </td>
                    </TableCell>
                    <TableCell
                      key="7"
                    >
                      <td
                        className=""
                        role="gridcell"
                      >
                        unavailable
                      </td>
                    </TableCell>
                  </tr>
                </TableRow>
              </tbody>
            </table>
          </Table>
        </MainTable>
      </TableList>
    </Provider>
  </Router>
</MemoryRouter>
`;
